openwisp_monitoring/monitoring/base/models.py

```diff
@@ -138,6 +138,9 @@ def check_threshold(self, value, time=None):
             notification_type = 'threshold_crossed'
         # ok: returned within threshold limit
         elif not crossed and self.is_healthy in [False, None]:
+            # First metric write and within threshold, do not raise alert
+            if self.is_healthy is None:
+                notify = False
```

Comment:
this would still trigger the sending of the `threshold_crossed` signal, which will lead to inconsistent behavior:

software consuming that signal will manifest the same issue we're seeing now with alerts being sent unnecessarily.

I would just return after `save()`.

Reply:
The signal `threshold_crossed` and it's receiver are being used in both situations (this can be improved). The [receiver](https://github.com/openwisp/openwisp-monitoring/blob/master/openwisp_monitoring/device/base/models.py#L233) of this signal also takes care of updating the status for a new device (unknown --> ok). This should be done by a separate method, I think. Without that, device health status might never get updated as can be seen from this [travis build](https://travis-ci.org/github/openwisp/openwisp-monitoring/jobs/698984432#L1047).

**Proposal**: Lets's have a separate `threshold_recovery` signal and separate the code in `threshold_crossed` receiver which is taking care of updating status during recoveries currently.